### PR TITLE
[7.x] Make the ArrayStore behaviour consistent with other cache stores

### PR DESCRIPTION
**Pull request by spawnia**
## Rationale

The `ArrayStore` is currently the only store that behaves fundamentally different from all other stores in two ways:

### Values are not (un-)serialized

The first is that values are not serialized/unserialized when storing them. That causes issues when switching out the cache driver. Other cache stores have to serialize to be able to store PHP objects.

For example, we had a bug in the serialization of a class, which we did not spot in our unit tests that used `ArrayStore`. It popped up once the `RedisStore` we use in production actually had called `serialize()`.

### Increment/Decrement does not respect TTL

As of now, incrementing an expired value in the `ArrayStore` does indeed increment and return a new value. At the same time, calling `get()` on the same key would return `null`.

This could not be the case in other cache stores, as they don't have the ability to "cheat" expire time in the same way the `ArrayStore` does.

## Functional Changes

The `ArrayStore` now calla `serialize` on every `put` and `unserialize` on every `get`.

Incrementing an expired value in the `ArrayStore` is now treated as if the value did not exist in the first place.

## Impact

I would categorize this as a bugfix, given that i expect implementations of an interface (in this case `Store`), to exhibit consistent observable behaviour.

Since the `ArrayStore` is commonly used for unit testing, it is especially important for users to have it behave the same as stores used for production purposes.

There are some cases where code might break:
- unserializabe objects can no longer be stored
- there is a reliance on incrementing expired values

I think both should be rare, probably mistakes, and are already the current behaviour of all other stores.

## Alternatives

If the existing behaviour of storing values without serialization has a valid use case, we could add a new store that does so.

Having a drop-in replacement that can be used for testing is really valuable to me.

### CODE REVIEW

File: tests/Cache/SerializableChecker.php [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    public function serialize()
+    {
+        return serialize('');
+    }
+
+    public function unserialize($serialized)
+    {
+        $this->wasSerialized = true;
+    }
+}
```
Comment (driesvints) [inline tests/Cache/SerializableChecker.php]: Put this class in the same test file if it's not used elsewhere.

**Review by driesvints — DISMISSED** [review]

File: tests/Cache/SerializableChecker.php (hunk 1 above)
Comment (spawnia) [reply tests/Cache/SerializableChecker.php]: Good point, done!

**Comment by mfn** [comment]
Regarging the serialization:

To me the thing is: it's an `ArrayStore`, not a `UseForUnitTestAsReplacementForExistingStoreArrayStore`.

What I mean: whilst the serialization implementation detail might be important to catch, it's equally important we don't force the whole serialize/unserialize on every user.

You can perfectly use the current ArrayStore "in production" too for whatever reason (e.g. you only need to cache within a current request or within a worker) without having the need for the serialization step.

TL;DR: IMHO this should be implemented in a different store.

**Comment by spawnia** [comment]
> You can perfectly use the current ArrayStore "in production" too for whatever reason (e.g. you only need to cache within a current request or within a worker)

I don't think that the cache is meant for that. PHP already has a mechanism for storing the result of a computation while in a single request: variables. If that is not sufficient or ergonomic for some use cases, we have other abstractions such as Singletons, Registries.

The `ArrayStore` will be usable as before, barring the mentioned exceptions. When high performance is needed, i don't think that going through the cache is the best approach anyways. I really don't see who this is hurting.

The whole point of an abstraction like `Cache` is to enable switching out the implementations, as they are hidden behind an interface.

**Comment by mfn** [comment]
I understand your points, was just important to me to provide a PoV.

Btw, at this point, serialize/unserialize is still an implementation detail: it's not on the contract in any way (e.g. for a custom serialize implementations)

**Comment by taylorotwell** [comment]
I do agree with @mfn that serialization is a coincidental implementation detail of the caches. It may be the case in the future that other drivers do not actually use serialization or use some other mechanism for storing objects. I also agree that people may be using this as a global single-request cache store in their production applications.

**Comment by mfn** [comment]
@spawnia a different approach was accepted via https://github.com/laravel/framework/pull/31295
